spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala

```diff
@@ -0,0 +1,283 @@
+      j -= 1
+      i -= 1
+    } else {
+      val gradOutputTable = gradOutput.toTable
+      require(gradOutputTable.length() == outputs.length,
+        "gradient tensor number is not equal to outputs node number")
+      while(j >= 0) {
+        outputs(j).element.updateGradInput(inputsBP(i), gradOutputTable[Tensor[T]](j + 1))
+        j -= 1
+        i -= 1
+      }
+    }
+
+    while(i >= 0) {
+      val curNode = executions(i)
+      var curGradOutput : Tensor[T] = null
+      curNode.nextNodes.foreach(n => {
+        val nextGradOutput = if (n.prevNodes.length == 1) {
+          n.element.gradInput.toTensor
+        } else {
+          val nextGradOutputTable = n.element.gradInput.toTable
+          nextGradOutputTable[Tensor[T]](n.prevNodes.indexOf(curNode) + 1)
+        }
+
+        if (curGradOutput == null) {
+          curGradOutput = nextGradOutput
+        } else {
+          curGradOutput.add(nextGradOutput)
+        }
+      })
+      curNode.element.updateGradInput(inputsBP(i), curGradOutput)
+      i -= 1
+    }
+
+    gradInput = if (inputs.length == 1) {
+      inputs(0).element.gradInput
+    } else {
+      seqToTable(inputs.map(_.element.gradInput))
+    }
+
+    gradInput
+  }
+
+  override def accGradParameters(input: Activity, gradOutput: Activity, scale: Double): Unit = {
+    var j = outputs.length - 1
+    var i = executions.length - 1
+    if (outputs.length == 1) {
+      // Expect a tensor here
+      outputs(j).element.accGradParameters(inputsBP(i), gradOutput.toTensor)
+      j -= 1
+      i -= 1
+    } else {
+      val gradOutputTable = gradOutput.toTable
+      require(gradOutputTable.length() == outputs.length,
+        "gradient tensor number is not equal to outputs node number")
+      while(j >= 0) {
+        outputs(j).element.accGradParameters(inputsBP(i), gradOutputTable[Tensor[T]](j + 1))
+        j -= 1
+        i -= 1
+      }
+    }
+
+    while(i >= 0) {
+      val curNode = executions(i)
+      var curGradOutput : Tensor[T] = null
+      curNode.nextNodes.foreach(n => {
+        val nextGradOutput = if (n.prevNodes.length == 1) {
+          n.element.gradInput.toTensor
+        } else {
+          val nextGradOutputTable = n.element.gradInput.toTable
+          nextGradOutputTable[Tensor[T]](n.prevNodes.indexOf(curNode) + 1)
+        }
+
+        if (curGradOutput == null) {
+          curGradOutput = nextGradOutput
+        } else {
+          curGradOutput.add(nextGradOutput)
+        }
+      })
+      curNode.element.accGradParameters(inputsBP(i), curGradOutput)
+      i -= 1
+    }
+  }
+
+  // Add a dummy output node, to get an one end graph. So the nodes that are not dependent by
+  // the outputs will be excluded
+  private val dummyOutput = new ModuleNode[T](null)
+  outputs.foreach(_ -> dummyOutput)
+  private val backGraph = dummyOutput.graph(reverse = true)
+
+  // Build execution plan
+  private val executions = backGraph.topologySort.filter(_.element != null).reverse
```

Comment:
Yes. But it's easy to write some following code if we exclude it. e.g. if you don't filter here, you have to filter in the next line: modules.appendAll. It's one time work, don't need to worry the perf